src/Microsoft.AspNet.OData.Shared/EdmStructuredObject.cs

```diff
@@ -224,8 +224,14 @@ internal static object GetDefaultValue(IEdmTypeReference propertyType)
                 if (propertyType.IsPrimitive() ||
                     (isCollection && propertyType.AsCollection().ElementType().IsPrimitive()))
                 {
+                    bool hasDefaultConstructor = (!clrType.IsClass) || clrType.GetConstructors(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static).
```

Comment:
Is it valid to include `BindingFlags.NonPublic` flag given that it'd resolve constructors (`private`, `protected` or `internal`) that we may not be able to invoke externally?
Is it also valid to include `BindingFlags.Static` given that the constructor is called automatically before the first instance is created or any static members are referenced?

I think you we should target the single parameter-less, public, non-static constructor. The following should do the job:
```csharp
bool hasDefaultConstructor = (!clrType.IsClass) || clrType.GetConstructor(Array.Empty<Type>());
```
See https://docs.microsoft.com/dotnet/api/system.type.getconstructor#system-type-getconstructor(system-type())
equivalent (I believe) to:
```
bool hasDefaultConstructor = (!clrType.IsClass) || clrType.GetConstructors(System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance).Any(x => x.GetParameters().Length == 0);
```
If you choose to use the second variant, please note that `GetParameters` method returns an array. You can use the `Length` property instead of the more expensive `Count` method.